### ISSUE TITLE
Suggesting to remove GENERATE_BUILD_REQUIREMENTS in meta.yaml

### ISSUE THREAD

**Issue by bobleesj** [Billingegroup/staged-recipes-cookiecutter#13]
## Problem

conda-forge recipe repo CI fails.

## Potential cause

In our current `meta.yaml` file, we have two `build` sections:

```
**build:**
  preserve_egg_dir: true
  noarch: python
  number: 0
  script: python -m pip install --no-deps --ignore-installed .

requirements:
  **build:**
    - GENERATE_BUILD_REQUIREMENTS
```

The `GENERATE_BUILD_REQUIREMENTS` section renders into:

```
**build:**
  preserve_egg_dir: true
  noarch: python
  number: 0
  script: python -m pip install --no-deps --ignore-installed .

requirements:
  **build:**
    - python >=3.10
    - setuptools >=62.0
```

I believe the second build is causing the `diffpy.snmf` staged recipe CI to fail as of September 1: https://github.com/conda-forge/staged-recipes/pull/27265

## Solution

We should remove the following section:

```
requirements:
  **build:**
    - python >=3.10
    - setuptools >=62.0
```

### Test case 1: My own package

I tested this with my Python package: https://github.com/conda-forge/staged-recipes/pull/27408

Initially, I used the `meta.yaml` from the `staged-recipes-cookicutter` repo, which caused the CI to fail. I then cross-checked with  other repos where CI passes. I then removed the second build section. Now the CI passes.

### Test case 2: `diffpy.utils`

In our current `diffpy.utils` https://github.com/conda-forge/diffpy.utils-feedstock/blob/main/recipe/meta.yaml, the CI passes and there's no extra section for building. 

@sbillinge

**Comment by sbillinge** [comment]
Good work!   I think we want to pin the versions though, especially for setup tools.  Is there a way to structure the meta.yaml to do that?  Is this problem coming from the cookicutter, or does this package deviate from the cookiecutter for some reason?

**Comment by sbillinge** [comment]
Great!  It would be good if we could capture some of this background information on the cookicutter readme for the future. 

I trust you to make a good recommendation how to structure this the best and I will merge it, but also some of this helpful background info copy posted into the README for future reference. 

The cookicutter meta.yaml was working with other packages.  So we know why it was paying before but not now?

**Comment by bobleesj** [comment]
> The cookicutter meta.yaml was working with other packages. So we know why it was paying before but not now?

Could you please confirm this?

I am seeing no conda-forge recipe PR being merged this year for diffpy packages:

https://github.com/conda-forge/staged-recipes/pulls?q=is%3Apr+is%3Aclosed+diffpy

Among the ones that are open for PR now, the tests are failing:

Ref: https://github.com/conda-forge/staged-recipes/pulls?q=is%3Apr+diffpy+is%3Aopen

For non-diffpy project, `bg-mpl-stylesheets`, it was merged a few years ago:

https://github.com/conda-forge/staged-recipes/pull/15860

**Comment by sbillinge** [comment]
The ones that are merged are already feedstocks.   Pdfmorph, utils etc.

**Comment by sbillinge** [comment]
I am not exactly sure if the tests are the same for staged recipes and for feedstocks, so that may be something to look at, but I imagine that they should be the same, or similar

**Comment by bobleesj** [comment]
> I am not exactly sure if the tests are the same for staged recipes and for feedstocks, so that may be something to look at, but I imagine that they should be the same, or similar

I think it's the same CI because once it's the recipe reppo is merged then the conda package is available.

"After merging the [PR](https://conda-forge.org/docs/glossary/#pr), our [CI](https://conda-forge.org/docs/glossary/#ci) infrastructure will build the package and make it available in the conda-channe"

REF: https://conda-forge.org/docs/maintainer/adding_pkgs/#feedback-and-revision


> The ones that are merged are already feedstocks. Pdfmorph, utils etc.

It doesn't seem like they are created with the current cookiecutter recipe repo because they do not have the 2nd build section in `meta.yaml`.

`bg-mpl-stylesheets`:
```
build:
  noarch: python
  number: 0
  script: python -m pip install --no-deps --ignore-installed .

requirements:
  host:
    - python >=3.10
    - setuptools
    - pip

  run:
    - python >=3.10
    - setuptools
    - matplotlib-base
```
- https://github.com/conda-forge/bg-mpl-stylesheets-feedstock/blob/main/recipe/meta.yaml

`diffpy.utils-feedstock`:

```
build:
  preserve_egg_dir: true
  noarch: python
  number: 2
  script: python -m pip install --no-deps --ignore-installed .

requirements:
  host:
    - python >=3.10
    - setuptools
    - pip

  run:
    - python >=3.10
    - setuptools
    - numpy >=1.3
```
- https://github.com/conda-forge/diffpy.utils-feedstock/blob/main/recipe/meta.yaml#L13-L28

**Comment by bobleesj** [comment]
To confirm - `bg-mpl-stylesheets` feedstock repo was created three years ago, and no second build section was added.

https://github.com/conda-forge/bg-mpl-stylesheets-feedstock/blob/06d708cb0020f5e4672f29e8ca63282706e950c8/recipe/meta.yaml

**Comment by sbillinge** [comment]
Good sleuthing work!  Let's maybe ask @Sparks29032 if he can remember how things came to be this way.  He built the conda-forge cookiecutter.  There is a reason that we pinned the setuptools version for sure so we may want to update those older meta.yamls but also we may have inadvertently introduced a bug. But @Sparks29032 is probably best to ask.

**Comment by Sparks29032** [comment]
I set the default for build requirements to these two packages, but realized they are not necessary and even duplicated for most packages. I can set the build default to blank, but I would rather keep that section under requirements as it is more clear how to maintain in the future if we introduce build dependencies to any package.

**Comment by Sparks29032** [comment]
Never mind, after checking it seems the build requirements are blank... this change was not added due to the cookiecutter. Someone likely manually added this and it is not a problem with the cookiecutter itself.

**Comment by Sparks29032** [comment]
I am against removing GENERATE_BUILD_REQUIREMENTS though as certain packages will require them. The default is to keep it blank.

**Comment by sbillinge** [comment]
So a question for @bobleesj is why does removing the build from requirements cause it to pass CI?  I don't think these are duplicate sections.  `build:` tells how to build (what script to run and so on) and 
```
requirements:
    build:
```
simply lists dependencies for the build to be successful.  So they are not duplicating anything.   But we do know that diffpy.snmf was failing CI, and then when the second build section was added, it passed.  So what is going on?

**Comment by Sparks29032** [comment]
Removing the python from the second build requirements seems sufficient, we do not need to remove the entire section.

**Comment by bobleesj** [comment]
> So a question for @bobleesj is why does removing the build from requirements cause it to pass CI? I don't think these are duplicate sections. `build:` tells how to build (what script to run and so on) and
> 
> ```
> requirements:
>     build:
> ```
> 
> simply lists dependencies for the build to be successful. So they are not duplicating anything. But we do know that diffpy.snmf was failing CI, and then when the second build section was added, it passed. So what is going on?


The current cookiecutter implementation automatically generates:

```
requirements:
  **build:**
    - python >=3.10
    - setuptools >=62.0
 ```

These `python` and `setuptools` do not belong under `build:`. Rather, they belong under `host:`. Please see the above comments: https://github.com/Billingegroup/staged-recipes-cookiecutter/issues/13#issuecomment-2325170535 and also the new info in the README.md.

> Removing the python from the second build requirements seems sufficient, we do not need to remove the entire section. The noarch python in the first build already adds python to the build env proper.

This will solve the issue.

**Comment by bobleesj** [comment]
@Sparks29032  To keep me informed for maintenance, do you know which diffpy packages require separate packages under `requirements/builds` in `meta.yaml`?

**Comment by Sparks29032** [comment]
@bobleesj Please update your version of this package. I have downloaded the current version of this cookiecutter and it makes an empty build section.

**Comment by bobleesj** [comment]
@Sparks29032  Thanks. Works with the latest version.

```
requirements:
  build:

  host:
    - python >=3.10
    - setuptools
    - pip

  run:
    - python >=3.10
    - setuptools
    - numpy
```

For projects that do not enter packages under `build `, can we remove the 

```
  build:

```

section?

**Comment by sbillinge** [comment]
@bobleesj can we close this issue?
